examples/vision/siamese.py

```diff
@@ -0,0 +1,232 @@
+introduced by Bromley and LeCun to solve signature verification as an image
+matching problem is now widely used for finding the level of similarity between
+two images. This algorithm evolved to have many different forms, although
+the one we are going to use is Duplet Network. A duplet network involves
+two sister networks for finding embeddings/encodings and then a distance
+heuristic to find out how much these embeddings differ from each other.
+
+"""
+
+"""
+## Necessary imports
+"""
+
+import random
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras.models import Model
+from tensorflow.keras.layers.experimental.preprocessing import Rescaling
+from tensorflow.keras.layers import Flatten, Dense, Concatenate, Lambda, Input
+from tensorflow.keras.layers import Conv2D, Activation,AveragePooling2D
+from tensorflow.keras.datasets import mnist
+from tensorflow.keras.optimizers import RMSprop
+from tensorflow.keras import utils
+import matplotlib.pyplot as plt
+from keras import backend as K
+
+"""
+## Load data
+"""
+
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
+
+# We change data type to float
+x_train = x_train.astype('float32')
+x_test = x_test.astype('float32')
+
+"""
+## Create pairs of images
+
+We want our model to be able to differentiate 0 from 1,2,3, .., 9 to do so we
+should teach it how each digit differs from each of the other digit, for this
+purpose we would select N random images from class A (lets say class of
+digit 0) and would pair it with N random images from class B (lets say class of
+digit 1), and would repeat this process untill class of digit 9. In this way we
+successfuly pair 0 with each other digit, now we would do the same for class of
+1,2,3 upto that of 9.
+"""
+
+def make_pairs(x, y):
+    num_classes = max(y) + 1
+    digit_indices = [np.where(y == i)[0] for i in range(num_classes)]
+
+    pairs = []
+    labels = []
+
+    for idx1 in range(len(x)):
+        # add a matching example
+        x1 = x[idx1]
+        label1 = y[idx1]
+        idx2 = random.choice(digit_indices[label1])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [1]
+    
+        # add a not matching example
+        label2 = random.randint(0, num_classes-1)
+        while label2 == label1:
+            label2 = random.randint(0, num_classes-1)
+
+        idx2 = random.choice(digit_indices[label2])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [0]
+
+    return np.array(pairs), np.array(labels).astype('float32')
+
+pairs_train, labels_train = make_pairs(x_train, y_train)
+pairs_test, labels_test = make_pairs(x_test, y_test)
+
+"""
+## Create dataset
+"""
+
+# 0 index of pairs_train contain one image of pair
+# and at index 1 the other image of pair is present
+# we can split entire vector of pairs into two havles,
+# later we can generate tf.data.Dataset using these halves
+# we will do the same for labels
```

Comment:
Suggestion:

- Refactor (grammar/syntax)
- Fix typo

```suggestion
# Index 0 of `pairs_train` contains one image of a pair,
# and index 1 - the other image of a pair.
# We can split the entire vector of pairs into two halves
# and later generate `tf.data.Dataset` objects using these halves
# We will repeat the process for labels.
```

(also, is it "image pairs" or "image of a pair"?)

Reply:
I made it more explicity in new commit.